src/ir/auth_logic/souffle_emitter.h

```diff
@@ -0,0 +1,143 @@
+    // there are no duplicate delcarations (which would otherwise happen
+    // whenever a predicate is referenced more than once with different
+    // arguments).
+    // To prevent instances of negated and non-negated uses of the predicate 
+    // from generating two declarations, the sign here is always positive.
+    Predicate PredToDeclaration(Predicate predicate) {
+      std::vector<std::string> decl_args;
+      for( int i=0; i < predicate.args().size(); i++ ) {
+        decl_args.push_back("x" + std::to_string(i));
+      }
+      return Predicate(predicate.name(), predicate.args(), kPositive);
+    }
+
+    std::string EmitVecSeparated(std::vector<std::string> vec,
+        std::string sep) {
+      std::string ret = "";
+      bool is_first = true;
+      for(auto elm : vec) {
+        if( !is_first ) { ret += sep; } else { is_first = false; }
+        ret += elm;
+      }
+      return ret;
+    }
+
+    std::string EmitPredicate(Predicate predicate) {
+      // Whenever a new predicate is encountered, it is added to the set of 
+      // declarations (which does not include duplicates because it is a set).
+      declarations_.insert(PredToDeclaration(predicate));
+      return (predicate.sign() == kNegated ? "!" : "")
+          + predicate.name()
+          + "(" + EmitVecSeparated(predicate.args(), ", ") + ")";
+    }
+
+    std::string EmitAssertionInner(Predicate predicate) {
+      return EmitPredicate(predicate) + ".";
+    }
+
+    std::string EmitAssertionInner(DLIRCondAssertion cond_assertion) {
+      std::string lhs = EmitPredicate(cond_assertion.lhs());
+      std::vector<std::string> rhs_translated;
+      for(auto arg: cond_assertion.rhs()) {
+        rhs_translated.push_back(EmitPredicate(arg));
+      }
+      std::string rhs = EmitVecSeparated(rhs_translated, ", ");
+      return std::move(lhs) + " :- " + std::move(rhs) + ".";
+    }
+
+    std::string EmitAssertion(DLIRAssertion assertion) {
+      return std::visit(
+          [this](auto value) { return EmitAssertionInner(value); },
+          assertion.GetValue());
+    }
+
+    std::string EmitProgramBody(DLIRProgram program) {
+      std::vector<std::string> body_translated;
+      for(auto assertion : program.assertions()) {
+        body_translated.push_back(EmitAssertion(assertion));
+      }
+      return EmitVecSeparated(body_translated, "\n");
+    }
+
+    std::string EmitOutputs(DLIRProgram program) {
+      std::string ret;
+      for (auto out : program.outputs()) {
+        ret += ".output " + out + "\n";
+      }
+      return ret;
+    }
+
+    std::string EmitDeclaration(Predicate predicate) {
+      std::string args = "";
+      bool is_first = true;
+      for(auto elm : predicate.args()) {
+        if(!is_first) { args += ", "; } else { is_first = false; }
+        args += elm + ": symbol";
+      }
+      return ".decl " + predicate.name() + "("
+        + args + ")";
+    }
+
```

Comment:
This looks like another place where `absl::StrJoin` would be a better choice.

Reply:
fixed